libcasr/src/stacktrace.rs

```diff
@@ -287,11 +288,226 @@ impl Filter for Stacktrace {
+    };
+    let (left, right) = (shift + pos, shift + pos + 2 * l - 1);
+    let interval_len = right - left + 1;
+    if intervals
+        .iter()
+        .filter(|x| !(x.1 < left || right < x.0))
+        .all(|x| x.2 < interval_len)
+    {
+        intervals.retain(|x| x.1 < left || right < x.0);
+        intervals.push((left, right, right - left + 1));
+    }
+}
+
+/// Perform Main-Lorentz algorithm
+///
+/// # Arguments
+///
+/// * `s` - given sequence
+///
+/// * `shift` - shift
+///
+/// * `intervals` - the resulting vector consisting of tuples such as (interval start, interval end, period length)
+///
+/// * `nae` - element that not equal to others
+///
+/// More about Main-Lorentz algorithm https://cp-algorithms.com/string/main_lorentz.html
+fn main_lorentz<T: PartialEq>(
+    s: &[&T],
+    shift: usize,
+    intervals: &mut Vec<(usize, usize, usize)>,
+    nae: &T,
+) {
+    let n = s.len();
+    if n < 2 {
+        return;
+    }
+
+    let len_u = n / 2;
+    let len_v = n - len_u;
+    let (u, v): (Vec<&T>, Vec<&T>) = (s[..len_u].to_vec(), s[len_u..].to_vec());
+    let (mut ru, mut rv) = (u.clone(), v.clone());
+    ru.reverse();
+    rv.reverse();
+    main_lorentz(&u, shift, intervals, nae);
+    main_lorentz(&v, shift + len_u, intervals, nae);
+
+    let pr1 = prefix_function(&ru);
+    let pr2 = prefix_function(&[v.clone(), vec![nae], u].concat());
+    let pr3 = prefix_function(&[ru, vec![nae], rv].concat());
+    let pr4 = prefix_function(&v);
+    for cntr in 0..n {
+        let (l, k1, k2) = if cntr < len_u {
+            (
+                len_u - cntr,
+                pr1.get(len_u - cntr).unwrap_or(&0),
+                pr2.get(len_v + 1 + cntr).unwrap_or(&0),
+            )
+        } else {
+            (
+                cntr - len_u + 1,
+                pr3.get(2 * len_u + len_v - cntr).unwrap_or(&0),
+                pr4.get(cntr - len_u + 1).unwrap_or(&0),
+            )
+        };
+        if k1 + k2 >= l {
+            add_interval(intervals, shift, cntr < len_u, cntr, l, *k1, *k2);
+        }
+    }
+}
+
+/// Return resulting vector with repeating intervals
+///
+/// # Argumnets
+///
+/// * `arr` - given sequence
+///
+/// * `not_an_element` - element that not equal to others
+///
+/// # Return value
+///
+/// Resulting vector with repeating intervals: (start index, end index, period length)
+fn get_interval_repetitions<T: PartialEq>(
+    arr: &[T],
+    not_an_element: &T,
+) -> Result<Vec<(usize, usize, usize)>> {
+    if let Some(idx) = arr.iter().position(|x| x == not_an_element) {
+        return Err(Error::Casr(format!(
+            "not_an_element is equal to element of given sequence at index {idx}"
+        )));
+    }
+    let links = arr.iter().collect::<Vec<&T>>();
+    let mut result = Vec::new();
+    main_lorentz(&links, 0, &mut result, not_an_element);
+
+    /// Calculate the minimal period of the given sequence.
+    fn get_period<T: PartialEq>(seq: &[&T]) -> usize {
```

Comment:
I cannot find this function here: https://cp-algorithms.com/string/main_lorentz.html

Reply:
it is my own function